crates/ruff_server/src/lint.rs

```diff
@@ -201,33 +200,94 @@ fn to_lsp_diagnostic(
 
     let data = (fix.is_some() || noqa_edit.is_some())
         .then(|| {
-            serde_json::to_value(&AssociatedDiagnosticData {
+            let edits = fix
+                .as_ref()
+                .into_iter()
+                .flat_map(Fix::edits)
+                .map(|edit| lsp_types::TextEdit {
+                    range: if let Some(notebook_index) =
+                        source_kind.as_ipy_notebook().map(Notebook::index)
+                    {
+                        edit.range()
+                            .to_notebook_range(
+                                source_kind.source_code(),
+                                index,
+                                notebook_index,
+                                encoding,
+                            )
+                            .1
+                    } else {
+                        edit.range()
+                            .to_range(source_kind.source_code(), index, encoding)
```

Comment:
I also need to refactor this logic: it's used multiple times and isn't easy to read.